data-products/src/main/scala/org/sunbird/analytics/model/report/ETBMetricsModel.scala

```diff
@@ -62,29 +78,65 @@ object ETBMetricsModel extends IBatchModelTemplate[Empty,Empty,FinalOutput,Final
       val reportConfig = JSONUtils.deserialize[ReportConfig](JSONUtils.serialize(configMap))
 
       val etbTextBookReport = events.map(report => {
-        if(report.etb.size!=0) report.etb.get else ETBTextbookReport("","","","","","","","","",0,0,0,0,0,"")
+        if(null != report.etb && report.etb.size!=0) report.etb.get else ETBTextbookReport("","","","","","","","","",0,0,0,0,0,"")
       }).filter(textbook=> !textbook.identifier.isEmpty)
 
       val dceTextBookReport = events.map(report => {
-        if(report.dce.size!=0) report.dce.get else DCETextbookReport("","","","","","","","",0,0,0,0,0,"")
+        if(null != report.dce && report.dce.size!=0) report.dce.get else DCETextbookReport("","","","","","","","",0,0,0,0,0,"")
+      }).filter(textbook=> !textbook.identifier.isEmpty)
+
+      val etbDialcodeReport = events.map(report => {
+        if(null != report.etbDialcode && report.etbDialcode.size!=0) report.etbDialcode.get else ETBDialcodeReport("","","","","","","","","","","","","","",0,0,"")
+      }).filter(textbook => !textbook.identifier.isEmpty)
+
+      val dceDialcodeReport = events.map(report => {
+        if(null != report.dceDialcode && report.dceDialcode.isDefined) report.dceDialcode.get else DCEDialcodeReport("","","","","","","","","","","","",0,"","")
       }).filter(textbook=> !textbook.identifier.isEmpty)
 
+      val scansDF = getScanCounts(config)
+
       reportConfig.output.map { f =>
         val etbDf = etbTextBookReport.toDF()
         CourseUtils.postDataToBlob(etbDf,f,config)
 
         val dceDf = dceTextBookReport.toDF()
         CourseUtils.postDataToBlob(dceDf,f,config)
+
+        val dialdceDF = dceDialcodeReport.toDF()
+        val dialcodeDCE= dialdceDF.join(scansDF,dialdceDF.col("dialcode")===scansDF.col("dialcodes"),"left_outer").drop("dialcode","noOfScans")
+        CourseUtils.postDataToBlob(dialcodeDCE,f,config)
+
+        val dialetbDF = etbDialcodeReport.toDF()
+        val dialcodeETB= dialetbDF.join(scansDF,dialetbDF.col("dialcode")===scansDF.col("dialcodes"),"left_outer").drop("dialcode","noOfScans")
+        CourseUtils.postDataToBlob(dialcodeETB,f,config)
       }
     }
     events
   }
 
-  def generateReports(config: Map[String, AnyRef])(implicit sc: SparkContext): (RDD[FinalOutput]) = {
+  def getScanCounts(config: Map[String, AnyRef])(implicit sc: SparkContext, fc: FrameworkContext): DataFrame = {
+    implicit val sqlContext = new SQLContext(sc)
+
+    val conf = config.get("etbFileConfig").get.asInstanceOf[Map[String, AnyRef]]
+    val bucket = conf("bucket")
+    val key = AppConf.getConfig("azure_storage_key")
+    val file = conf("file")
+    val url = s"wasb://$bucket@$key.blob.core.windows.net/$file"
+    val finalUrl = if("local".equals(config("store"))) conf("filePath").asInstanceOf[String] else url
```

Comment:
This should be cloud agnostic..it should support local file, s3 and azure

Reply:
Done.